File: benchmarks/binary_trees.rb
class Node
  attr_reader :left, :right
  def initialize(left, right)
    @left = left
    @right = right
  end
end

def build(depth)
  if depth == 0
    Node.new(nil, nil)
  else
    Node.new(build(depth - 1), build(depth - 1))
  end
end

def check(node)
  if node.nil?
    0
  else
    1 + check(node.left) + check(node.right)
  end
end

t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
total = 0
50.times do
  t = build(16)
  total += check(t)
end
t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)

puts total
puts "bench_ns=#{t1 - t0}"

File: benchmarks/binary_trees.js
function build(depth) {
    if (depth === 0) {
        return { left: null, right: null };
    }
    return { left: build(depth - 1), right: build(depth - 1) };
}

function check(node) {
    if (node === null) return 0;
    return 1 + check(node.left) + check(node.right);
}

const t0 = process.hrtime.bigint();
let total = 0;
for (let i = 0; i < 50; i++) {
    const t = build(16);
    total += check(t);
}
const t1 = process.hrtime.bigint();

console.log(total);
console.log(`bench_ns=${t1 - t0}`);
